### DISCUSSION TITLE
Review before/after feature behavior

### DISCUSSION THREAD

**Discussion by gasparnagy** [reqnroll/Reqnroll#124] [Ideas]
It seems that the current before/after feature behavior has some strange aspects, so this discussion can be used to decide on what we should do with this feature.

Here is the current behavior (discussed at #119, but reviewed).

Note: I will refer often to the term **test-thread** below. As the name is a bit misleading, here is it's definition: Reqnroll somehow assigns the test execution requests to **test-threads**. These are basically logical flows (not physical threads!) and Reqnroll ensures that the test execution within the same test-thread are not started parallel. (How it ensures that is out of scope here, and it might depend on the test runner framework, but this is given.) Of course in an unlucky situation it can happen that every scenario will be executed in a unique test-thread, so there is no guarantee that is is "optimal", but usually it is. 

The `ITestRunner` interface is the representation of the execution methods of a particular test-thread, so we have as many `TestRunner` instances as the number of test-threads.

1. Currently the hooks are triggered by the class-level setup/teardown methods of the test runners (MsTest, xUnit, NUnit). These methods are static (or have static behavior).

2. For Reqnroll to behave correctly, the before/after feature hook **must be run on the same "test-thread" as the related scenario execution** and hooks. (Otherwise some stange null ref exceptions come.)

3. As the test runners might use different tactics to run the class-level setup/teardown methods and the tests, especially in parallel situations, it is difficult to ensure that Reqnroll assigns these executions to the same test-thread. The way is slightly different for each test runner, but mainly it works in a way that **during the class-level setup method it saves the `ITestRunner` to a static field** and this is reused for the test execution.

4. MsTest has a special handling of the class-level teardown method, because it might be called _delayed_, so a scenario of the next class might come faster to the test-thread than the teardown of the previous one in some situations. To handle this, there is a **special trick** implemented in `TestExecutionEngine.OnFeatureStartAsync`: if the previous one was not finished, we finish that first.

So altogether the behavior is pretty complex and it depends heavily on the behavior of the test runners (that might even change). 

There are some important consequences of the fact that Reqnroll requires the before/after feature hook to be run on the same test-thread as the related scenarios:

- The before/after feature hooks of a concrete feature might be executed multiple times (if the test runner randomizes the scenarios across feature files).
- When the tests are running parallel, the before/after feature hooks of a concrete feature might be executed parallel to each-other (e.g. if a scenario from that feature comes in in multiple test-threads)
- The current behavior on the other hand ensures that before the execution of any scenario, the before feature hooks have been called and that the after feature hooks will be called after.

Originally, the before/after feature hooks have been designed for caching some automation infrastructure elements that are expensive to initialize and related to a particular feature file (not easy to find a good example though). 

With the modern test runners (parallel, randomized order), this benefit is quite limited, because usually the scenarios of the same feature file are anyway not running in a sequence block. But of course this depends on the runner configuration as well.

//CC @Code-Grump @obligaron

**Comment by gasparnagy** [comment]
My goal would be that we remove the dependency of the test runner's (MsTest, xUnit, NUnit) behavior for the feature hooks. This could be achieved in the following way:
- We remove (obsolete) the `OnFeatureStartAsync`, `OnFeatureEndAsync` methods from `ITestRunner`
- We pass in the `FeatureInfo` together with the `ScenarioInfo` to the runner in the `OnScenarioInitialize` method.
- Reqnroll itself will detect if the "current feature" has changed on the current test-thread and fires the before/after feature hooks
- At the end, when all test execution finished (just before the AfterTestRun hook), we invoke the after feature hooks of all "remaining" open test-threads (this depends on the fix of #123). 
- Optionally we can have some idle detection for the test-threads and if there is no further scenario execution started for a test-thread for some time, we close the test-thread and run the related after feature hook (even earlier then AfterTestRun). This might be useful for parallel execution, because usually at the end only a few longer test runs.

My other idea would be to deprecate this feature. The behavior can be simulated with before/after scenario anyway if necessary.

What do you think?

**Reply by Code-Grump** [reply]
I would _really_ like to change our nomenclature to not use the word "thread", as it's definitely easy to think it _is_ relating to execution threads (an already confusing area.) I believe I understand it to be equivalent to an execution context, but even then I'm not certain of a test-thread's lifecycle. I will avoid using the word "thread" as much as possible to reduce ambiguity.

A clear explanation of what "test-thread" encapsulates (either here or in a separate discussion) would be appreciated.

---

I haven't personally seen a practical use for feature-level hooks, but they are analogous to class-level hooks prevalent in all major test-runners, so I have to believe they're useful to somebody. Perhaps they'll be useful for us if we wanted to create a project on-disk and then run a number of scenarios before cleaning up the project at the end of the test.

The complexity in this discussion is definitely concurrency, where the feature level blocks need to maintain guarantees of:
- Before Feature runs **once** for a given feature before **any** of the feature's scenarios are run, but after the Before Test Run hooks
- After Feature runs **once** for a given feature after **all** the feature's scenarios have run, but before the After Test Run hooks
No guarantees can (or should) be made about _exactly_ when these things will happen, just that they happen relative to other important events.

I don't believe the idea of "changing feature" really doesn't make sense in this model, where all features could execute in parallel, in serial, or some interpolated mixture. Additionally, the end-user is free to run one test, or all the tests, or all the tests containing the letter 'z'. We're coupled to the execution engine of whatever test-framework we run on top-of, so we really need to be flexible enough to handle the full spectrum that the end-user might want to configure.

Importantly, the context created at the feature level needs to flow into the scenarios. Instances registered with the feature-scoped container should be available for all the scenarios via their containers (whilst still maintaining a separation between the scenario-scoped services.)

My thoughts run counter to yours @gasparnagy where I think the answer lies in _tighter_ integration with the test-framework, relying on their established hooks to run at the times when they're expected to, and using appropriate mechanisms to flow our execution context between those stages. In some runners that might be as simple as leveraging `AsyncLocal` or `ExecutionContext`, in others we'll have to grab their native contexts and propagate that way.

The test-runners I'm familiar with already make the guarantees we need about when hooks run in relation to other important hooks, so re-engineering that seems like we're not taking advantage of that established and relied-on work. I may have a naive view of the problem and not understand why it's difficult, but in my mind I feel like we "only" need to propagate a context object through the test-runner and call its methods at the appropriate times.

**Comment by gasparnagy** [comment]
> I would really like to change our nomenclature to not use the word "thread", as it's definitely easy to think it is relating to execution threads (an already confusing area.) I believe I understand it to be equivalent to an execution context, but even then I'm not certain of a test-thread's lifecycle.

I understand your concern and I am fine finding another name, but "execution context" is a term used for async and it is not exactly the same as ours (although it is similar) so that would be confusing as well. So if you have a good suggestion please share it. I will also think on that. Unfortunately it is also in the API (like `TestThreadContext` class), so renaming is a bit complicated, but not impossible.

> I haven't personally seen a practical use for feature-level hooks, but they are analogous to class-level hooks prevalent in all major test-runners, so I have to believe they're useful to somebody.

Code based test tools, you can declare and use fields, so there the class-level hooks can be used to initialize/manipulate these. In Gherkin files you cannot declare variables and the steps are not tightened to a particular feature file, so therefore the usage is more limited and the analogy does not work fully.

> Before Feature runs **once** for a given feature before any of the feature's scenarios are run, but after the Before Test Run hooks

Unfortunately this is not enough. The example usages of the before/after feature hooks that I have seen (before parallel times) was that they 
* updated the database with some records in before feature in order to provide the necessary base data for the scenarios
* logged in the user in the browser in before feature, because all scenarios needed a logged in user

I'm not saying that these are good usages, but I have seen people using it this way. For these cases it is not enough if the before feature runs some time (but not necessarily directly) before the first scenario in that feature file, but they need to be run _directly_ before the first scenario.

If we want to apply a more relaxed specification (with the **once** before), we should find a good use-case where this is useful for otherwise we just break the current one without adding a benefit. And this is where I usually get stuck. I can't really find a good use-case for before/after feature that makes sense in modern execution environments (parallel, random order).

> I don't believe the idea of "changing feature" really doesn't make sense in this model, where all features could execute in parallel, in serial, or some interpolated mixture. [...] I think the answer lies in tighter integration with the test-framework, relying on their established hooks to run at the times when they're expected to, and using appropriate mechanisms to flow our execution context between those stages.

I also don't have a strong opinion here.

But I think tighter integration with the test-framework is what we have now. We rely on the test-framework infra, [applied the workarounds for the unmatching behavior or MsTest](https://github.com/reqnroll/Reqnroll/blob/0aecb3480ce0ff8cd41eee34a8c592c92ebb3956/Reqnroll/UnitTestProvider/IUnitTestRuntimeProvider.cs#L8) and need to store the test runner in a static field to ensure that the feature context and the scenario context are linked (they need to run on the same test runner instance). 

So the question is that if we keep the tighter integration with the test-framework, how could we improve the current situation? Do you have some ideas?

**Reply by Code-Grump** [reply]
> For these cases it is not enough if the before feature runs some time (but not necessarily directly) before the first scenario in that feature file, but they need to be run directly before the first scenario.

That's a really crucial difference that comes down more to the test execution model: if the caller needs to log in, then run their scenarios before any other feature's scenarios run, **they categorically cannot run their features in parallel.** They'd need to be able to run their features in a linear fashion, something they can probably achieve via their test-framework. Their needs are likely to be compatible with the guarantees I listed.

If the issue is one of time-sensitivity, that will almost entirely come down to execution time. We could encourage it to run closer to the scenarios, but a slow machine will still break this. All the test frameworks I'm used to run their before-class hooks as the very last thing before running the first test on that class. The _after_ hooks are much harder to pin down, with some test-frameworks waiting until the end of all tests before performing those calls.

This is one of those areas where I would prefer to encourage users to choose a suitable test-framework for their needs. I genuinely think one of the **best** features of Reqnroll is its ability to run on top of a variety of test-frameworks and take advantage of the specific quirks each one offers.

> So the question is that if we keep the tighter integration with the test-framework, how could we improve the current situation?

I think we improve things by relaxing the requirements and making test-thread focused on just flowing context between hooks and only guarantee relative sequencing. I do believe there are feature-scoped activities that make sense, generally infrastructure things where we'd like to set up some state, within the context of the scenarios we're about to run, such as creating a file or setting up an isolated database - these are all things that can scope to a feature. If the tests run concurrently, that means the author has to be aware of concurrency and can't assume it's all going to work by some kind of magic. We can guarantee their setup and teardown will happen "around" the scenarios and that should be sufficient.

This would remove the burden on us to try and maintain "workarounds". I believe we should be vocal that we run when the test-frameworks tell us to run, and that if those moments aren't what the end-user needs (i.e. if they were writing a unit test, it would not work for them) then it suggest they should either stop doing what they're doing or choose a test-framework that does meet that need.

**Reply by obligaron** [reply]
I like your idea of looking at the use cases for the before/after feature hooks.

> - updated the database with some records in before feature in order to provide the necessary base data for the scenarios
> - logged in the user in the browser in before feature, because all scenarios needed a logged in user

With Reqnroll and the `FeatureContext` this is possible, if the `BeforeFeature` looks something like this:

```csharp
[BeforeFeature]
public static void BeforeFeature(ITestRunner testRunner)
{
    var newUserId = Guid.NewGuid();
    MyDatabase.AddNewUser(newUserId);
    testRunner.FeatureContext.Add("UserId", newUserId);
}
```

So you could have multiple features running in parallel and they would still work.

If some kind of global synchronisation is needed `BeforeTestRun` with a static is usable.

All of this (running features or scenarios in parallel) requires help from the Reqnroll user. He needs to understand what impact running features or scenarios in parallel might have on his own code.
Reqnroll should help him with hooks, but we can't help him with his own logic.
Finally, the user has the choice of which behavior to choose (no parallelism, feature level parallelism or scenario level parallelism) and which test framework to use. So he can choose what is best for his use case. 🙂

Regarding backwards compatibility:
For example my PR doesn't change the current behavior (expect that some ITestRunner instances aren't reused) and still supports scenario-level parallelization. Why? Because the user has to opt-in to use `ExecutionScope.MethodLevel`. If he still uses the `ExecutionScope.ClassLevel`  or no parallelization, nothing changes for him.

> But I think tighter integration with the test-framework is what we have now. We rely on the test-framework infra, [applied the workarounds for the unmatching behavior or MsTest](https://github.com/reqnroll/Reqnroll/blob/0aecb3480ce0ff8cd41eee34a8c592c92ebb3956/Reqnroll/UnitTestProvider/IUnitTestRuntimeProvider.cs#L8) and need to store the test runner in a static field to ensure that the feature context and the scenario context are linked (they need to run on the same test runner instance).

Interesting, I didn't know about this special. I'll have to check it out later. Thanks for highlighting it.🙂

**Reply by gasparnagy** [reply]
@Code-Grump @obligaron Thx for your insights. 

I think many of these questions is actually coming down to the understanding of the test-thread concept (let's put the naming problem aside for a second). So let me come back to that first, before answering your comments.

I suggest you should look at the https://github.com/reqnroll/Sample-ReqOverflow/blob/main/ReqOverflow.Specs.WebUI, because that shows this concept and through that you can better understand this concept than from my words (and maybe can better explain it to others).

That sample runs the tests via Chrome, using Selenium. Starting up a Chrome instance (even in headless mode) takes quite much time, therefore with this setup, it is not so easy to optimize the execution with "usual" methods, because:
1. You can run the tests in a single thread, caching the chrome driver (in a static field for example), but then it takes too much time
2. You can run the tests parallel, but then you need to create a new chrome driver for every scenario.
3. If you use class-level parallelization only, you can cache the chrome driver per feature, but this is still not ideal if you have many feature files with a few scenarios only (the overhead of creation is still big) or if you have feature files with many scenarios (harder to parallelize).

The linked sample uses a 4th option for this: It utilizes the test-thread concept of Reqnroll, to create only as many chrome browsers as many parallel "threads" (workers) are used by the test runner to execute the tests. Let's say you set the level of parallelization (worker count) to 4, that means you will have 4 browser instance created. Reqnroll provides you with a `TestThreadContext` class for each of these workers, so you can use that to cache your browser. You can see this code in https://github.com/reqnroll/Sample-ReqOverflow/blob/main/ReqOverflow.Specs.WebUI/Support/BrowserFactory.cs#L44.

How Reqnroll figures out when to create a `TestThreadContext` is a technical detail, but Reqnroll gets this information from the test runner framework as much as possible (e.g. NUnit simply provides `TestContext.WorkerId` that contains a unique id for each worker). The different strategies are implemented by the test framework providers via the `IUnitTestGeneratorProvider.GetTestWorkerIdExpression` method.

All scenario executions that share the same `TestThreadContext` run in sequence, because this is how we defined `TestThreadContext`. 

So basically we have the following groupings for the scenario executions:

- Test Run (contains all scenario executions), also called the "global context" - there is only one of this, coordinated by the singleton `TestRunnerManager` class
  - Test Thread (one per worker), also called test-thread context - there are multiple of these, but the scenarios run in sequence within that. They are controlled by the `ITestRunner` interface (and the related `TestRunner` and `TestExecutionEngine` class)
    - Feature context - wrap the scenarios within a test-thread. You can have multiple Feature context for the same feature if their scenarios are divided to multiple workers.
      - Scenario context - the inner atomic block. In case of scenario outlines, every example has its own scenario context.

Is it OK up to this point? (@Code-Grump Would "worker" be a better name?)

**Reply by Code-Grump** [reply]
I follow the idea, thank you very much for the illustrative example! 😀 I've encountered this exact problem with Selenium drivers before and solved it in a completely different way: I created a resource pool and each scenario would request a driver instance from the pool. If there were none in the pool, we'd create a new one. When a scenario had finished executing, it would be released back to the pool for a future scenario to pick up.

At the time, it never even occurred to me that SpecFlow had a way to bind a resource to a "worker" (yes I like that name much better!) so it wasn't something I had considered as a potential solution. 

Knowing what I do now, I still think I prefer having a pool to solve this, as it doesn't require coupling the resource-management to any execution-model's specifics, and whilst I can see it being convenient to have this kind of "floating" resource scope, it seems like a complex beast to tame, both to understand its usage and to implement successfully.

**Reply by gasparnagy** [reply]
Visualization:
![image](https://github.com/reqnroll/Reqnroll/assets/144701/7b41ce99-4cc3-4b63-b431-67d5a766db91)

**Reply by ajeckmans** [reply]
It is also confusing, because whilst reading the reply from @gasparnagy my mental model first came up with Test Thread and Feature Context being inverted to what it is now. Might be just me, but I'd avoid that altogether and use a pool as well. I've actually used a singleton pool in the past for this. 
I think the semantics are a little bit different though.

**Reply by gasparnagy** [reply]
> "worker" (yes I like that name much better!)

OK. Let's consider this then as a naming improvement.

> Knowing what I do now, I still think I prefer having a pool to solve this, as it doesn't require coupling the resource-management to any execution-model's specifics

Yes. That is a good point, and I can agree. Actually the test-thread (worker) concept was needed by Reqnroll (SpecFlow) anyway to be able to serialize some of the infrastructure, so the possibility of using the same for resource management was just a useful side-effect. 

As proper pooling is not that easy to implement, this might be convenient for simpler projects. As many users came to Reqnroll (SpecFlow) from the "synchronous, single threaded, cache in static fields" type of usages and projects, it is good IMHO if we can provide an easy solution that replaces the static fields for parallel. I think even a before/after "worker" hook would make sense, to further simplify these use cases.

BTW: Since xUnit does not provide any info about its worker assignment in any form, for xUnit the `TestThreadContext` is actually solved by a simple pooling (see [here](https://github.com/reqnroll/Reqnroll/blob/main/Reqnroll.Generator/UnitTestProvider/XUnit2TestGeneratorProvider.cs#L417) and [here](https://github.com/reqnroll/Reqnroll/blob/0aecb3480ce0ff8cd41eee34a8c592c92ebb3956/Plugins/Reqnroll.xUnit.ReqnrollPlugin/XUnitParallelWorkerTracker.cs#L7)).

I will continue following up on the specific feature hook details in the evening or tomorrow.

**Reply by gasparnagy** [reply]
> my mental model first came up with Test Thread and Feature Context being inverted to what it is now

You are not alone with this, I guess. But it the other relation (global - feature - test-thread (worker) - scenario) would be even more problematic from the implementation point of view, because we could not maintain a strict parent-child relationship between the contexts. 

> Might be just me, but I'd avoid that altogether and use a pool as well. I've actually used a singleton pool in the past for this.

I can just answer the same as I did to @Code-Grump. It is absolutely fine, if you use other sorts of ways to manage your resources. And the same applies to the feature context and the feature hooks - this can be also implemented with a hash table indexed by the feature name. 

But we should not forget that you (we) are probably power-users of Reqnroll, so not necessarily representing the every-day usage. I do training and coaching for BDD, so I see usages in the field. How much we should simplify the things at the price of making magic black box solutions is a good question of course though (that I keep hesitating about all the time). My vision was always that SpecFlow (Reqnroll) is a tool only. It should not dictate the usage, but provide enough (safe) help for the teams to improve themselves. So providing "training wheels" here and there is not necessarily bad as long as it does not block the advanced usages.

**Reply by Code-Grump** [reply]
In the sprit of making things easier for users, perhaps the most "friendly" thing we could do is looking at implementing an easier resource-management system with a "pooled" lifecycle type?
- Singleton / Scoped
- Transient
- Pooled

A pool implementation isn't hard to write with "lease" semantics and we get to bake that into our DI system transparently, since we know when a consumer's lifecycle has ended and the leased object can be released back to the pool.

---

At some point I'd like to have a look at whether we still actually need the test-thread for _our_ implementation. If we can remove our dependency on it, it might make our own execution internals easier to work with and opens the way for considering obsoleting it in favour of concepts that are less-coupled to an execution model. There's definitely a cognitive load with having to think about execution models and resource-lifetimes together that I would personally like to remove.

**Reply by gasparnagy** [reply]
> At some point I'd like to have a look at whether we still actually need the test-thread for our implementation.

Yes. I will make a review of this.

**Reply by gasparnagy** [reply]
> At some point I'd like to have a look at whether we still actually need the test-thread for our implementation.

I did the analysis now. 

As far as I see, for our implementation we really *need* this only to ensure the behavior of the before/after feature hooks. So we are circling back to our original problem. I think I have a better understanding now, so let me write it down before I forget. 

It became pretty long at the end, so made it a top level comment, see below.

**Reply by obligaron** [reply]
I tried to replace the workaround for MsTest in https://github.com/reqnroll/Reqnroll/pull/128.

Thanks @gasparnagy for the explanation. Now I understand more what a "test-thread" is. 🙂

If we don't guarantee what "test-thread"/`TestThreadContext` a scenario/`TestRunner` is using, we could use pooling on our side (like the xUnit runner). The only difference to the current behavior would be that a `TestThreadContext` can change when running different tests (we could also remove this by reusing the `TestThreadContext` from the feature level, if it's not used yet).

**Comment by gasparnagy** [comment]
Yet another try to describe the before/after feature hook problem after some analysis.

Let's start with the specification of the before/after feature hooks: **The before/after feature hooks are _directly_ before and after a set of scenarios that are placed and only placed in the same feature file. The hooks can use a storage (FeatureContext) that has the lifetime between the pairing before and after feature hook.** A _consequence_ of this is, that the hooks of a particular feature might be executed multiple times (and therefore have multiple FeatureContext) if the scenarios of that feature file are not executed in a single block.

This specification requires a local sequence (in order to identify the "set of scenarios that are placed and only placed in the same feature file") and that requires the test-thread (worker) concept at the end.

The technical implementation of the feature hooks currently:

In order to trigger the before/after feature hooks we used the class-level setup/teardown infrastructure of the test execution frameworks, but this works differently for all 3 frameworks (and probably also based on what parallel model you use with them). 

One of the key challenges here is how we connect the data that the user saved in the before feature hook to the execution of the scenario. So far (at least for xUnit and MsTest) we used **static fields** to store this data (by saving the TestRunner into a static field - and TestRunner is unique per test-thread (worker)). (The MsTest "ClassInitialize" method has to be static, so there are not much other options we had.)

This technical implementation was good as long as the parallelization was per-class at max, because the test executions of the same class could happily work together on the same static TestRunner. This is obviously not true for test-level parallelization.

Side note: The PR #119 of @obligaron uses a separate static field for storing the TestRunner for the before/after feature hook, but this is not sufficient, because is might happen that the data you save into the FeatureContext in the before feature hook is not "visible" during the scenario execution.

Maybe too early to conclude here, but I can see the following possible options:

1) We keep the before/after feature hook spec (and hence the test-thread (worker) concept), but do **not** trigger these from the test execution framework class-level setup/teardown infrastructure, but doing it "on demand" during test execution. With that we remove the need to pass data between the class initialize method and the test method.

2) We change the before/after feature hook spec in a way that it does not offer a direct before/after execution and it runs the before/after feature hooks only once per feature regardless of how the scenarios are scheduled. This way we would not need to pair the feature hooks with the scenario executions as we can simply store the FeatureContext in a has by the class name. We could also eliminate the test-thread (worker) concept as well.

3) We deprecate (and eventually delete) the before/after hook and all the complexity caused by that (including test-thread (worker) concept). This can be combined temporarily with option 1) to be able to support scenario-level parallelism. 

---

I feel that many of you would tend to go for option 2), and I don't have any problem with that in principle. As far as I understood your main arguments are basically making the Reqnroll codebase and the concepts simpler. I can agree to this, but this is really a tough decision. Not from the technical point of view, but from project & product perspective. My fears about this are the following:

* This is a breaking change that is hard to communicate. Code that has been written in before/after feature hooks will compile fine, but suddenly it works differently. (That is the worst kind of breaking change.)
* I have seen usages of before/after feature hooks that were relying on the direct before/after execution (so there is an existing need), but I still cannot see any concrete example for the modified specs. And I am a bit skeptical for changing to something that we haven't even seen to be useful.
* This is a LOT of work because all Reqnroll infrastructure is heavily built on the test-thread (worker) concept.
* I'm not sure we want to spend our time *now* on that (we have quite many open topics and a big demand on the living doc generator)


If we want to have simplicity, I would rather go for option 3). That is a clearer situation and - as you have mentioned around the resource management - we can provide examples of pooling, etc. where the same behavior can be achieved without feature hooks.

The option 1) can be used to have a quick win: it simplifies the code gen, works with scenario-level parallelism, easy and is not a breaking change. But it keeps the complexity and the "cognitive load" of the test-thread (worker) concept. This can also be seen as a "delaying" option as we can later still go for option 2) or 3).

**Reply by obligaron** [reply]
I tried to replace the workaround for MsTest in https://github.com/reqnroll/Reqnroll/pull/128.
Hopefully the AfterFeature hook should now be called (nearly) after the scenario is finished.

> Side note: The PR https://github.com/reqnroll/Reqnroll/pull/119 of @obligaron uses a separate static field for storing the TestRunner for the before/after feature hook, but this is not sufficient, because is might happen that the data you save into the FeatureContext in the before feature hook is not "visible" during the scenario execution.

The PR should carry on data that was set in `BeforeFeature` to the `FeatureContext` to the scenario execution. At least I added this to my test. But perhaps I missed something, so I would be happy if you could point me to a direction. 🙂 
Side-note: Perhaps in the PR, to keep the discussion here more focused.

**Reply by Code-Grump** [reply]
I would love to go for option 2 to make the code based simpler for all involved, but I think all the reasons not to do it now are valid. I believe breaking changes need to be shipped in a major version, so this is probably something we can leave in the backlog for a future time. 

Option 1 feels like a change we could make happen quickly and helps improve the situation. 

I would like to spend some time looking at alternative mechanisms to hooks, in the hope we can make them simpler, encourage them as an alternative/soft-deprecation of hooks and reduce the impact of breaking changes to hooks. But after the Roslyn generator 😅

**Comment by Code-Grump** [comment]
This issue seems relevant https://github.com/reqnroll/Reqnroll/issues/189 to the discussion.
